mlx/ops.cpp

```diff
@@ -3777,6 +3777,123 @@ array block_masked_mm(
   return out;
 }
 
+/** Compute matrix product with matrix-level gather */
+array block_sparse_mm(
+    array a,
+    array b,
+    std::optional<array> lhs_indices_ /* = std::nullopt */,
+    std::optional<array> rhs_indices_ /* = std::nullopt */,
+    StreamOrDevice s /* = {} */) {
+  // If no indices, fall back to full matmul
+  if (!lhs_indices_ && !rhs_indices_) {
+    return matmul(a, b, s);
+  }
+
+  // Do shape checks for operands
+  int in_a_ndim = a.ndim();
+  int in_b_ndim = b.ndim();
+
+  if (a.ndim() == 0 || b.ndim() == 0) {
+    throw std::invalid_argument(
+        "[block_sparse_mm] Got 0 dimension input. Inputs must "
+        "have at least one dimension.");
+  }
+
+  if (a.ndim() == 1) {
+    // Insert a singleton dim in the beginning
+    a = reshape(a, {1, -1}, s);
+  }
+  if (b.ndim() == 1) {
+    // Insert a singleton dim at the end
+    b = reshape(b, {-1, 1}, s);
+  }
+
+  if (a.shape(-1) != b.shape(-2)) {
+    std::ostringstream msg;
+    msg << "[block_sparse_mm] Last dimension of first input with shape "
+        << a.shape() << " must match second to last dimension of"
+        << " second input with shape " << b.shape() << ".";
+    throw std::invalid_argument(msg.str());
+  }
+
+  // Type promotion
+  auto out_type = result_type(a, b);
+  if (!issubdtype(out_type, floating)) {
+    std::ostringstream msg;
+    msg << "[block_sparse_mm] Only real floating point types are supported but "
+        << a.dtype() << " and " << b.dtype()
+        << " were provided which results in " << out_type
+        << ", which is not a real floating point type.";
+    throw std::invalid_argument(msg.str());
+  }
+
+  a = astype(a, out_type, s);
+  b = astype(b, out_type, s);
+
+  // Handle broadcasting
+  std::vector<int> bsx_a(a.shape().begin(), a.shape().end() - 2);
+  std::vector<int> bsx_b(b.shape().begin(), b.shape().end() - 2);
+
+  int n_batch_a = 1;
+  int n_batch_b = 1;
+  for (auto& i : bsx_a)
+    n_batch_a *= i;
+  for (auto& i : bsx_b)
+    n_batch_b *= i;
+
+  // Pull and broadcast indices
+  array lhs_indices =
+      lhs_indices_.value_or(reshape(arange(n_batch_a, uint32, s), bsx_a, s));
+  array rhs_indices =
+      rhs_indices_.value_or(reshape(arange(n_batch_b, uint32, s), bsx_b, s));
```

Comment:
I'm not a big fan of this style with optionals because I think it will always evaluate the operand which can be kind of inefficient to construct all those arrays and never use them if you specify `lhs_indices`.  Maybe cleaner to do it with a lambda like so:
```c++
auto indices_or_default = [](const std::optional<array>& indices, std::vector<int> bsx_shape) {
  if (indices.has_value()) {
    return indices.value();
  } else {
     int n_batch = 1;
     for (auto& i : bsx_shape)
       n_batch *= i;
    return reshape(arange(n_batch, uint32, s),  bsx, s);
  }
}
```